includes/core/ValueObjects/API/Request/class-create-charge.php

```diff
@@ -0,0 +1,111 @@
+<?php
+/**
+ * Create charge request.
+ *
+ * @package WooCommerce\Payments
+ */
+
+namespace WCPay\Core\ValueObjects\API\Request;
+
+use WCPay\Core\Enums\Http_Methods;
+use WCPay\Core\Enums\Wc_Pay_Endpoints;
+
+/**
+ * Create charge request.
+ */
+final class Create_Charge extends Base_Request {
```

Comment:
Create_Charge should cover all possible charge requests towards WC pay server. Allowing extending the class allows overwriting some core functionalities like validation and request preparing. In this case, it is better to create a brand-new request using base class as foundation.

Do you see benefit in making it non-final? I just wanted to make strict rules in building requests towards wcpay client.

Reply:
On one hand I like the idea of `final` classes, on the other hand I'd like to see how this would play in real life.

[This line](https://github.com/Automattic/woocommerce-payments/blob/dev/wc-pay-request-class/includes/class-wc-payment-gateway-wcpay.php#L975) is a prime example of consumer code within WCPay, which should be isolated, and not live where it lives right now. Also, there's the `save_payment_method_to_platform` paramter for `create_and_confirm_intention`, which seems like something additional. With this class marked as `final`, I'm not sure how WooPay is expected to add the `woopay_has_subscription` paramter? 

> In this case, it is better to create a brand-new request using base class as foundation.

Are they expected to copy-paste this whole class within their codebase, rename it, and add the necessary parameters?

The main objective of the issue this PR resolves is such cases, and it seems we've left those behind. While most requests are likely better to implement in other PRs/issues, I'd like this PR to include:

1. An example of standard requests, used by WCPay, as it already does. However, I want to see those requests in action, not just in theory/standalone directory. You have already started the implementation of `create_intention`, but `create_and_confirm_intention` is used much more heavily through the plugin. Let's focus on that request, and do it in a way, where you can go to the checkout page, and complete a purchase through those new methods. Then we'll be able to discuss stuff [like this](https://github.com/Automattic/woocommerce-payments/blob/b2681db0b2d7e6096b50a672e78de60a6dc7f15a/includes/wc-payment-api/class-wc-payments-api-client.php#L2003-L2014), and decide what to do with that early (ex. a trait).
2. An example of a WooPay-specific parameter(s), isolated outside of the current `includes` folder, but still working. This means that somewhere in our code there should be a filter/action, which allows them to add those parameters safely, and this PR should include an example of how that is done. The example can then be converted to docs right before merging.

Reply:
We could add these extra fields or custom logic could be part of the request builders via functions. The good example of this function inside Create_Intention is `confirm_intention()` (probably I would have to rename this function since it could be a misleading)

For example:

This line of code could be replaced and implemented inside the Request class like this:

  Current code:
  
```
if ( '1' === $order->get_meta( '_woopay_has_subscription' ) ) {
         $additional_api_parameters['woopay_has_subscription'] = 'true';
}
```

Proposed approach using Request class

// Inside Create_Intention class, we could create additional methods to add small functionalities like this. One of the examples is to create something like this:

```
public function maybe_has_woopay_subscription( $order ){
   if ( '1' === $order->get_meta( '_woopay_has_subscription' ) ) {
         $this->additional_parameters['woopay_has_subscription'] = 'true';
    }
}
$intent = new Create_Intention();
$intent->maybe_has_woopay_subscription( $order )
```

Or something that could be useful by setting `WP_Order` in `Create_Intention`

```
//Inside Request class.

private $wp_order;

public function set_wp_order( $order ) {
    // ... will validate is order object ...
    $this->wp_order = $order;
}
public function maybe_has_woopay_subscription( ){
   if ( $this->wp_order && '1' === $order->get_meta( '_woopay_has_subscription' ) ) {
         $this->additional_parameters['woopay_has_subscription'] = 'true';
    }
}
$intent = new Create_Intention();
$intent->set_wp_order ( $order );
$intent->maybe_has_woopay_subscription() //Maybe add this automaticlly ?
```

But adding a bunch of these cases could be problematic because you don't know all of them, so a practical approach would be to remove final mark and allow other teams to extend the class, or they could add their own methods via PRs in the future.

I can fully understand your point now. Let me come up with some of the ideas. I will commit them with examples for easier understanding with a code examples I could find in our codebase.